File: CallbackTechniquesTests/KeyValueObserving/KeyValueObservingTests.m
//
//  KeyValueObservingTests.m
//  CallbackTechniques
//
//  Created by Alexey Naumov on 30/01/2017.
//  Copyright © 2017 Alexey Naumov. All rights reserved.
//

#import "KeyValueObserving.h"

@interface ObjCKeyValueObservingTests : XCTestCase

@property (nonatomic, strong) ObjCKeyValueObservingCallee *callee;
@property (nonatomic, strong) ObjCKeyValueObservingCaller *caller;

@end

@implementation ObjCKeyValueObservingTests

- (void)setUp {
  [super setUp];
  self.callee = [ObjCKeyValueObservingCallee new];
  self.caller = [ObjCKeyValueObservingCaller new];
  [self.callee startObservingObject:self.caller];
}

- (void)tearDown {
  [self.callee stopObservingObject:self.caller];
  [super tearDown];
}

- (void)testConnectivity {
  XCTAssertFalse(self.callee.wasCalled);
  [self.caller changeValue];
  XCTAssertTrue(self.callee.wasCalled);
}

- (void)testPerformance {
  [self measureBlock:^{
    repeat(number_of_iterations, [self.caller changeValue]);
  }];
}

@end

File: CallbackTechniquesTests/KeyValueObserving/KeyValueObservingTests.swift
//
//  KeyValueObservingTests.swift
//  CallbackTechniques
//
//  Created by Alexey Naumov on 30/01/2017.
//  Copyright © 2017 Alexey Naumov. All rights reserved.
//

class SwiftKeyValueObservingTests: XCTestCase {
  
  var callee: SwiftKeyValueObservingCallee!
  var caller: SwiftKeyValueObservingCaller!
  
  override func setUp() {
    super.setUp()
    callee = SwiftKeyValueObservingCallee()
    caller = SwiftKeyValueObservingCaller()
    callee.startObserving(object: caller)
  }
  
  override func tearDown() {
    callee.stopObserving(object: caller)
    super.tearDown()
  }
  
  func testConnectivity() {
    XCTAssertFalse(callee.wasCalled)
    caller.changeValue()
    XCTAssertTrue(callee.wasCalled)
  }
}
